Global Announcement Banner

CharlieSears opened the pull request:
Getting this PR draft started to ensure feedback can be delivered early. This is in response to feedback on: https://github.com/DefectDojo/django-DefectDojo/pull/5648

# Goal
Enable administrators to optionally show system-wide notifications to users.

Allow users to dismiss the banner

Functionality should be similar to Jira's implementation of an "announcement banner" that is visible on all pages.

# Decisions Made
Please feel free to contradict any of these decisions to ensure the best result.
- Announcement will be displayed beneath and beside navigation components but above page content. This was due to the use of hard-coded locations in sb-admin-2 for the top and left nav components.
- Announcement dismissal state should be stored in the database. This is similar (but inverted) to how alerts are handled now. A UI based approach with cookies or local storage would always have a minor delay that could be jarring to a user to constantly and dramatically pop the announcement in/out. Django should handle what is rendered
- Style should be select-able to influence the gravity of the announcement. Utilized built-in [alert-{style} from bootstrap](https://getbootstrap.com/docs/3.4/components/#alerts).

# Todo

- [x] Basic idea implemented
- [x] Dismiss-ability should come from Django, JS on page load is too slow and distracting
- [x] Integration tests

# UI
Please help me ensure all cases are covered.

Short messages messages
![image](https://user-images.githubusercontent.com/19413468/150585569-b9f2c9ab-0d5a-4c42-a32e-429bfb244a95.png)

Wrapping/long messages
![image](https://user-images.githubusercontent.com/19413468/150587720-962d3b15-aa43-4cc7-89ff-a683eb5826bb.png)

Correct sizing with minimized left nav
![image](https://user-images.githubusercontent.com/19413468/150586015-d43d7b70-544b-4fad-893e-56d8bdd78c38.png)

Correct sizing when "mobile"/small screen width
![image](https://user-images.githubusercontent.com/19413468/150586777-590e7b6e-fad1-4521-b75a-cdfb6ec13e7b.png)

Hidden with other 'global' elements on metrics dashboard
![image](https://user-images.githubusercontent.com/19413468/150587221-81020eaa-b74a-4041-8a3d-749e8b5f68c0.png)

devGregA commented:
@CharlieSears we like this PR. Are you ready to move it from draft to actual?

CharlieSears commented:
@devGregA sorry for the late reply. I'll get this up to snuff and a merge prepped. I don't believe I'll have it ready for 2.16.0 given how close that is.

CharlieSears commented:
I added translations since the initial state of this branch is so old. However there are massive changes in the en django.po file. I'm happy to revert those changes or include them as part of a separate PR if anyone prefers.

I simply ran `django-admin makemessages -l en` from the `./dojo` folder with `gettext@0.19.8.1-10build1

CharlieSears commented:
Happy with the functionality at this point. Going to get some tests thrown together and turn this into a proper PR from draft.

Maffooch commented:
Sounds great @CharlieSears !

Maffooch commented:
@CharlieSears just looked through everything since it was marked as ready and all looks great! No changes from me :)
I see you've got some empty checkboxes for unit/integration tests. Do you think those would be ready for the 2.17.0 release that is planned for December 7th? If not, I will bump the milestone up

CharlieSears commented:
> @CharlieSears just looked through everything since it was marked as ready and all looks great! No changes from me :) I see you've got some empty checkboxes for unit/integration tests. Do you think those would be ready for the 2.17.0 release that is planned for December 7th? If not, I will bump the milestone up

@Maffooch , unfortunately not. I'm in the middle of a move and won't have time to complete that. I'll defer to y'all as maintainers. I'm happy to merge this in without and add the tests later, but I'm not a fan of creating something without tests if I can help it :)

CharlieSears commented:
Added integration tests, removed unit test task due to minimal business logic which is fully covered by integration tests.

Happy to add them if yall see fit.

kiblik commented:
> ~I added translations since the initial state of this branch is so old. However there are massive changes in the en django.po file. I'm happy to revert those changes or include them as part of a separate PR if anyone prefers.~
> 
> ~I simply ran `django-admin makemessages -l en` from the `./dojo` folder with `gettext@0.19.8.1-10build1`~
> 
> Edit: Removed

Solution and prevention are coming: https://github.com/DefectDojo/django-DefectDojo/pull/7457

CharlieSears commented:
Going for longest lived branch here but I'm sure I'm not even close.

Thank you all for your patience and interest. Excited to get this in and in use.

Maffooch reviewed (approved):
@CharlieSears Great work here! Big thanks for sticking with one of the longest living branches in dojo history 😂 

@blakeaowens please review

blakeaowens commented on dojo/templates/dojo/dismiss_announcement.html:
```diff
@@ -0,0 +1,23 @@
+{% extends "base.html" %}
```
```suggestion
{% extends "base.html" %}
{% load i18n %}
```

Dismissing an announcement currently throws a 500 for me. Need to load internationalization from Django to use translation tag.

blakeaowens commented on dojo/templates/dojo/dismiss_announcement.html:
```diff
@@ -0,0 +1,23 @@
+        <form class="form-horizontal" method="post">
+            {% csrf_token %}
+            {{ form }}
+
```
```suggestion
<br />
```

Would be nice to have a break here to center the button within the "Danger Zone" block.

blakeaowens commented:
Besides my two suggested changes above, everything else looks great!

CharlieSears commented:
> Besides my two suggested changes above, everything else looks great!

Thank you! Not sure how I missed the internationalization tag, added both your suggestions and testing locally now with latest from `dev` merged.

csears-cvet commented:
@Maffooch / @blakeaowens , are either of you able to re-run tests?

Integration tests are passing locally, but getting intermittent failures on `test_test` and `findings_test` in the actions. A re-run usually solves things, but I have to wait for an update to `dev` so I can merge a change and avoid a change for the sake of a change. Plus I'd rather re-run intermittent failures than the whole suite to avoid heating the universe too much haha.

CharlieSears commented:
@Maffooch / @blakeaowens , are either of you able to re-run tests?

Integration tests are passing locally, but getting intermittent failures on test_test and findings_test in the actions. A re-run usually solves things, but I have to wait for an update to dev so I can merge a change and avoid a change for the sake of a change. Plus I'd rather re-run intermittent failures than the whole suite to avoid heating the universe too much haha.

Sorry for re-send
